### ISSUE TITLE
Custom field "JavaScript date" decrements by one day with each submission of the form where used

### ISSUE THREAD

**Issue by bkeplinger** [liedekef/events-made-easy#534]
I created custom fields "Start date" and "End date" for use in displaying non-event items (in this case, a staff listing). It works, but I noticed some peculiar behavior when modifying these fields. Any date I pick gets decremented by one day when I submit the form, so I need to select one day in the future for the date I want.

This also happens if I modify anything on the form but leave the dates untouched.

**Comment by bkeplinger** [comment]
First, I want to give a bit more information... I created HTML comments that output the Start date and End date as I submitted (and I display those fields in admin). They retain the date I originally submitted even though the fields display dates decremented by one day. For editing any person, opening the form displays dates decremented by one day, even though the correct dates display in admin.

When I created the fields, I chose "Date (Javascript)" and specified the format "Y-m-d", and the condition is to show those fields for all people.

**Comment by liedekef** [comment]
I can't reproduce this here (all dates stay with the value entered). Can it be the date on the server or your laptop is incorrect somehow? Can I have a demo?

**Comment by bkeplinger** [comment]
Here are two screenshots.
The main People admin view:
![Admin 1](https://github.com/liedekef/events-made-easy/assets/5461060/e5f068b3-7053-4de1-a146-c586aa0086d4)

A view of the form for this person immediately after I open it:
![Admin 2](https://github.com/liedekef/events-made-easy/assets/5461060/998ee0e3-e7a1-4e7a-8e08-ea3b73921280)

**Comment by bkeplinger** [comment]
I just tested with an event. Everything is fine. This only happens with the custom fields dates.

**Comment by liedekef** [comment]
Lke I said: I can't reproduce this here (all dates stay with the value entered). Can it be the date on the server or your laptop is incorrect somehow?
Also: check the html code for the custom field when editing a person (using "inspect" and see what the value is of the date entered. See this screenshot:
![image](https://github.com/liedekef/events-made-easy/assets/75667/e11b119a-0a7d-40ce-acbc-5a80991dca30)
there you can see the value entered is "2024-04-03" for the "real" value and for the value with format Y-m-d (which happens to be the same format in this case, but the logic still applies)

**Comment by bkeplinger** [comment]
For the screenshot of the form (where the selected date is 2022-04-29), this is inspect screenshot:
![Admin 3](https://github.com/liedekef/events-made-easy/assets/5461060/02689749-2214-4e53-b464-bf1ebcc45091)

The input value on the hidden field is 2022-04-29 while the input underneath that is 2022-04-30. This is not the case with yours, I see.

**Comment by liedekef** [comment]
This is very weird, as the code that generates this html is using the same value for both lines. See eme-formfields.php lines 943 and 944:
```
                        $html       = "<input type='hidden' name='$field_name' id='$field_name' value='$value' $class_att>";
                        $html      .= "<input $required_att readonly='readonly' $disabled type='text' name='dp_{$field_name}' id='dp_{$field_name}' data-date='$value' data-date-format='$dateformat' data-alt-field='#$field_name' class='eme_formfield_fdate $class'>";
```
So something else must be interfering here.
Does this still happen if you disable all other plugins?
Which EME version do you use?

**Comment by bkeplinger** [comment]
I'll look at that code tomorrow...

I update EME today: Version 2.4.41

I delete inactive plugins and deactivated all except EME. The result is the same.

**Comment by bkeplinger** [comment]
Had a couple minutes to check. Yes, the same. Very odd indeed.

**Comment by liedekef** [comment]
Check the time on your laptop
Try with a different browser
Try on another laptop

**Comment by bkeplinger** [comment]
The result in Edge is the same:
`<input type="hidden" name="dynamic_personfields[4][FIELD14]" id="dynamic_personfields[4][FIELD14]" value="2022-04-29" class="nodynamicupdates dynamicfield"><input readonly="readonly" type="text" name="dp_dynamic_personfields[4][FIELD14]" id="dp_dynamic_personfields[4][FIELD14]" data-date="2022-04-30" data-date-format="Y-m-d" data-alt-field="#dynamic_personfields[4][FIELD14]" class="eme_formfield_fdate nodynamicupdates dynamicfield">`

Setting the site up on another PC is something I'd like to avoid but I guess it's worth a shot.

Meanwhile, I added some debugging code:
`			$html       = "1: '$value'";
			$html      .= "<input type='hidden' name='$field_name' id='$field_name' value='$value' $class_att>";
			$html      .= "2: '$value'";
			$html      .= "<input $required_att readonly='readonly' $disabled type='text' name='dp_{$field_name}' id='dp_{$field_name}' data-date='$value' data-date-format='$dateformat' data-alt-field='#$field_name' class='eme_formfield_fdate $class'>";
			$html      .= "3: '$value'";
`

Here's the output I got:
![Admin 4](https://github.com/liedekef/events-made-easy/assets/5461060/d5cd9ae6-7459-4967-9fa6-d0b0cf1770a2)

And here is what inspect get me:
`1: '2022-04-30'<input type="hidden" name="dynamic_personfields[4][FIELD14]" id="dynamic_personfields[4][FIELD14]" value="2022-04-29" class="nodynamicupdates dynamicfield">2: '2022-04-30'<input readonly="readonly" type="text" name="dp_dynamic_personfields[4][FIELD14]" id="dp_dynamic_personfields[4][FIELD14]" data-date="2022-04-30" data-alt-field="#dynamic_personfields[4][FIELD14]" class="eme_formfield_fdate nodynamicupdates dynamicfield">3: '2022-04-30'`

**Comment by liedekef** [comment]
Your first inspect-screenshot had extra data-attributes (data-dashline-rid and data-form-type), your last no longer (and is more in line with the code I created). I'm guessing something is interfering here ... is there some kind of date-jquery plugin in your theme (or another plugin), or some kind of custom filter/hook? I know that you said you tried disabling all plugins, but what you've shown with your debug screenshot can only be explained by some other code changing things 
Are you even sure it is the EME calendar that pops up? This is what it looks like:
![image](https://github.com/liedekef/events-made-easy/assets/75667/586a9c6a-b25d-4746-ae1d-e032b0c9d550)

**Comment by liedekef** [comment]
Also (based on your screenshots), you've tried in firefox/chrome or so: in firefox apparently the custom date/time format was not set/used in the form display itself (while it is in chrome: a conflict with initialization I guess). This is now corrected for firefox too. See these changes in eme.js : https://github.com/liedekef/events-made-easy/commit/5cbfebbb196cc852e7fc1133709974eb79527feb#diff-a9f7818952e6a4b21bd984046def9fcbd35026d05a15e533172300c00ec1eddc

**Comment by liedekef** [comment]
Currently I can't reproduce this in any way, so unless I can get access to your setup I need to close this as non-reproducable ...

**Comment by bkeplinger** [comment]
I can't blame you. 

I was able to set this lite up on my laptop, and the same thing happens in Firefox and Edge (and Chrome on my PC, which I hadn't tested before). I created two new date fields, without spaces in the names, in case this may have contributed although I expected it wouldn't have, and it didn't. I didn't give the new Javascript field an HTML attribute. The result was the same. With the HTML field, it was correct, as expected.

I'm going to submit this in Stack Overflow. Maybe someone somewhere in the world has run into this vexing problem.

**Comment by liedekef** [comment]
Well, the fdatepicker.js code hasn't changed in 2 years, so that can't be it :-)
The changes to eme-formfields.php (rendering/doing the html) are not in that section either.
I suspect something is changing the date after the html is sent back to the browser, so something jquery-related (a javascript running). It could be EME related, but I can't reproduce this here at all.
Is it possible to have access to your setup?

**Comment by bkeplinger** [comment]
I probably just didn't notice, then... I do have access, laptop and PC (both Windows 10). It will be next week before I can do anything on this.

**Comment by bkepl** [comment]
Here are a couple more things it may be helpful for you to know. 

- I tried this on an event (using the default date fields) and everything is as it should be. 
- I cleared the cache in my browsers.

**Comment by liedekef** [comment]
I figured as much :-) No custom filters/hooks?

**Comment by bkeplinger** [comment]
The time is correct, but in any case, even if I change it, my PC will go out and sync itself to the correct time.

On a whim I changed the date format on my PC to Y-m-d format and as I expected, this had no effect.

I deleted both fields, then I added one only; same result. Then I wondered what would happen if I chose Datetime (Javascript). That kept the date as I entered it and didn't decrement on an update or successive updates. I changed the field to Date (Javascript) and the decrementing behavior returned. I changed it back to Datetime (Javascript) and no decrementing occurred.
